File: reference/CodeExamples/dotNet/VisualBasicExamples/Code/disconnectedobj/merging/MergeExample.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports Db4objects.Db4o
Imports Db4objects.Db4o.Config
Imports Db4objects.Db4o.Linq

Namespace Db4oDoc.Code.DisconnectedObj.Merging
    Public Class MergeExample
        Private Const DatabaseFileName As String = "database.db4o"


        Public Shared Sub Main(ByVal args As String())
            RunMergeExample()
        End Sub

        Private Shared Sub RunMergeExample()
            CleanUp()

            StoreCar()
            PrintCars()

            Dim car As Car = GetCarByName("Slow Car")
            UpdateCar(car)

            UpdateWithMerging(car)


            PrintCars()

            CleanUp()
        End Sub

        Private Shared Sub UpdateWithMerging(ByVal disconnectedCar As Car)
            ' #example: merging
            Using container As IObjectContainer = OpenDatabase()
                ' first get the object from the database
                Dim carInDb As Car = GetCarById(container, disconnectedCar.ObjectId)

                ' copy the value-objects (int, long, double, string etc)
                carInDb.Name = disconnectedCar.Name

                ' traverse into the references
                Dim pilotInDB As Pilot = carInDb.Pilot
                Dim disconnectedPilot As Pilot = disconnectedCar.Pilot

                ' check if the object is still the same
                If pilotInDB.ObjectId.Equals(disconnectedPilot.ObjectId) Then
                    ' if it is, copy the value-objects
                    pilotInDB.Name = disconnectedPilot.Name
                    pilotInDB.Points = disconnectedPilot.Points
                Else
                    ' otherwise replace the object
                    carInDb.Pilot = disconnectedPilot
                End If

                ' finally store the changes
                container.Store(pilotInDB)
                ' #end example
                container.Store(carInDb)
            End Using
        End Sub

        Private Shared Sub UpdateCar(ByVal car As Car)
            car.Name = "Fast Car"
            car.Pilot.Points = 300
        End Sub

        Private Shared Sub PrintCars()
            Using container As IObjectContainer = OpenDatabase()
                Dim cars As IList(Of Car) = container.Query(Of Car)()
                For Each car As Car In cars
                    Console.WriteLine(car)
                Next
            End Using
        End Sub

        Private Shared Function GetCarById(ByVal container As IObjectContainer, ByVal id As Guid) As Car
            Return (From car As Car In container _
                Where car.ObjectId.Equals(id) _
                Select car).Single()
        End Function

        Private Shared Function GetCarByName(ByVal carName As String) As Car
            Using container As IObjectContainer = OpenDatabase()
                Return (From car As Car In container _
                        Where car.Name.Equals(carName) _
                        Select car).First()
            End Using
        End Function


        Private Shared Sub StoreCar()
            Using container As IObjectContainer = OpenDatabase()
                container.Store(New Car(New Pilot("Joe", 200), "Slow Car"))
            End Using
        End Sub


        Private Shared Sub CleanUp()
            File.Delete(DatabaseFileName)
        End Sub


        Private Shared Function OpenDatabase() As IObjectContainer
            Dim configuration As IEmbeddedConfiguration = Db4oEmbedded.NewConfiguration()
            configuration.Common.ObjectClass(GetType(IDHolder)).ObjectField("guid").Indexed(True)
            Return Db4oEmbedded.OpenFile(configuration, DatabaseFileName)
        End Function
    End Class
End Namespace

File: reference/CodeExamples/dotNet/CSharpExamples/Code/disconnectedobj/merging/MergeExample.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Db4objects.Db4o;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Linq;

namespace Db4oDoc.Code.DisconnectedObj.Merging
{
    public class MergeExample
    {
        private const string DatabaseFileName = "database.db4o";


        public static void Main(string[] args)
        {
            RunMergeExample();
        }

        private static void RunMergeExample()
        {
            CleanUp();

            StoreCar();
            PrintCars();

            Car car = GetCarByName("Slow Car");
            UpdateCar(car);

            UpdateWithMerging(car);


            PrintCars();

            CleanUp();
        }

        private static void UpdateWithMerging(Car disconnectedCar)
        {
            // #example: merging
            using (IObjectContainer container = OpenDatabase())
            {
                // first get the object from the database
                Car carInDb = GetCarById(container, disconnectedCar.ObjectId);

                // copy the value-objects (int, long, double, string etc)
                carInDb.Name = disconnectedCar.Name;

                // traverse into the references
                Pilot pilotInDB = carInDb.Pilot;
                Pilot disconnectedPilot = disconnectedCar.Pilot;

                // check if the object is still the same
                if (pilotInDB.ObjectId.Equals(disconnectedPilot.ObjectId))
                {
                    // if it is, copy the value-objects
                    pilotInDB.Name = disconnectedPilot.Name;
                    pilotInDB.Points = disconnectedPilot.Points;
                }
                else
                {
                    // otherwise replace the object
                    carInDb.Pilot = disconnectedPilot;
                }

                // finally store the changes
                container.Store(pilotInDB);
                container.Store(carInDb);
                // #end example
            }
        }

        private static void UpdateCar(Car car)
        {
            car.Name = "Fast Car";
            car.Pilot.Points = 300;
        }

        private static void PrintCars()
        {
            using(IObjectContainer container = OpenDatabase()){
                IList<Car> cars = container.Query<Car>();
                foreach (Car car in cars)
                {
                    Console.WriteLine(car);
                }
            }
        }

        private static Car GetCarById(IObjectContainer container, Guid id)
        {
            return container.Query(delegate(Car car) { return car.ObjectId.Equals(id); })[0];
        }

        private static Car GetCarByName(string carName)
        {
            using (IObjectContainer container = OpenDatabase())
            {
                var result = (from Car car in container
                              where car.Name.Equals(carName)
                              select car).First();

                return result;
            }
        }


        private static void StoreCar()
        {
             using(IObjectContainer container = OpenDatabase())
             {
                 container.Store(new Car(new Pilot("Joe", 200), "Slow Car"));
             }
        }


        private static void CleanUp()
        {
            File.Delete(DatabaseFileName);
        }


        private static IObjectContainer OpenDatabase()
        {
            IEmbeddedConfiguration configuration = Db4oEmbedded.NewConfiguration();
            configuration.Common.ObjectClass(typeof (IDHolder)).ObjectField("guid").Indexed(true);
            return Db4oEmbedded.OpenFile(configuration, DatabaseFileName);
        }
    }
}